### PR TITLE
Statistics refactor

### PR DESCRIPTION
**Pull request by ilaydacansinkoc**
Current developments include such changes:

- Separation of CPU and Memory Charts.
- Average Memory and CPU Consumption in time.
- CPU and Memory consumption per user

Following Developments:

- List 3 Users that has most Memory and CPU consumption.

### CODE REVIEW

**Review by DSanyal19 — DISMISSED** [review]

File: src/pages/api/admin/statistics/getTableData.ts [hunk 1]
```diff
@@ -30,17 +45,33 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse<
...
+
+      for (const row of globalWorkspaceList.rows) {
+        workspaceListFromK8s.body.items &&
+          workspaceListFromK8s.body.items.forEach((each: any) => {
+            const name = each.metadata?.name;
+            if (name === row.name) {
+              workspaceListFromK8sArray.push({ name, user: row.userId });
+            }
+          });
+      }
```
Comment (larnon) [inline src/pages/api/admin/statistics/getTableData.ts]: No need for this part.

File: src/pages/api/admin/statistics/getTableData.ts [hunk 2]
```diff
@@ -30,17 +45,33 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse<
         user: 'admin',
       });
       await questdbClient.connect();
-      const sessionsData = await questdbClient.query(`SELECT * FROM '${DB_TABLE_NAMES.GLOBAL_SESSIONS}'`);
-      const workspacesData = await questdbClient.query(`SELECT * FROM '${DB_TABLE_NAMES.GLOBAL_WORKSPACES}'`);
-      const globalUsageData = await questdbClient.query(`SELECT * FROM '${DB_TABLE_NAMES.GLOBAL_USAGE}'`);
+      const globalWorkspaceList = await questdbClient.query(
+        `SELECT * FROM '${DB_TABLE_NAMES.GLOBAL_WORKSPACE_LIST}' WHERE isDeleted = false`
+      );
+      const workspaceListFromK8s = await k8s.getWorkspaceList();
```
Comment (larnon) [inline src/pages/api/admin/statistics/getTableData.ts:51]: This is the single source of truth for workspaces and it already comes with user information for each workspace.

File: src/pages/api/admin/statistics/index.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        const rows = [sessionList.rows, usageList.rows, workspaceList.rows, allWorkspaceList];
+        return res.status(200).json({ rows });
+      } catch (error) {
+        return res.status(400).json('Error getting statistics');
+      }
+    default:
+      res.setHeader('Allow', ['GET']);
+      res.status(405).end(`Method ${method} Not Allowed`);
+  }
+}
```
Comment (larnon) [inline src/pages/api/admin/statistics/index.ts]: No need for this API. We already had an API in getTableData.ts for getting all the global tables data in one fetch.

File: src/pages/admin/statistics.tsx [hunk 4]
```diff
@@ -5,13 +5,24 @@ import Collapsible from '../../components/Collapsible';
 import GlobalResourceUsageGraph from '../../components/Graphs/GlobalResourceUsageGraph';
 import GlobalWorkspacesGraph from '../../components/Graphs/GlobalWorkspacesGraph';
 import GlobalSessionsGraph from '../../components/Graphs/GlobalSessionsGraph';
+import annotationPlugin from 'chartjs-plugin-annotation';
```
Comment (larnon) [inline src/pages/admin/statistics.tsx]: This import is not used here.

File: src/components/Graphs/PerUserResourceUsageGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+import { DB_TABLE_NAMES, DB_TABLE_ROW_TYPES } from '../../../types/Database';
+import { toast } from 'react-toastify';
+import { useQuery } from '@tanstack/react-query';
+import { memo, useContext } from 'react';
+import { Context } from '../../context/Context';
+import annotationPlugin from 'chartjs-plugin-annotation';
+
+dayjs.extend(localizedFormat);
+
+const GlobalResourceUsageGraph = () => {
```
Comment (larnon) [inline src/components/Graphs/PerUserResourceUsageGraph.tsx]: Duplicate component name.

File: src/components/Graphs/PerUserResourceUsageGraph.tsx [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            ),
+          ],
+          datasets: getDataForCPUConsumptionPerUserChart(),
+        }}
+      />
+    </>
+  );
+};
+
+export default memo(GlobalResourceUsageGraph);
```
Comment (larnon) [inline src/components/Graphs/PerUserResourceUsageGraph.tsx]: Duplicate component name.

**Review by larnon — CHANGES_REQUESTED** [review]

File: src/pages/api/admin/statistics/getTableData.ts (hunk 2 above)
Comment (larnon) [inline src/pages/api/admin/statistics/getTableData.ts:51]: This is the single source of truth for workspaces and it already comes with user information for each workspace.

Comment (larnon) [inline src/pages/api/admin/statistics/getTableData.ts]: K8s API is the single source of truth for workspaces and it already comes with user information for each workspace.

**Review by larnon — CHANGES_REQUESTED** [review]

**Review by larnon — DISMISSED** [review]

**Review by larnon — APPROVED** [review]

**Review by sevvalekici — APPROVED** [review]

**Pull request merged**